### PR TITLE
Add `RelationError` error type to pageserver rather than string parsing error messages

### PR DESCRIPTION
**Pull request by awestover**
## Problem
Some places in pageserver relied on string parsing of error messages. A better solution is to have a custom error type.
## Summary of changes
Added a custom error type for `RelationError`
## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by awestover** [comment]
sorry I'll fix the code style

File: pageserver/src/pgdatadir_mapping.rs [hunk 1]
```diff
@@ -875,32 +885,38 @@ impl<'a> DatadirModification<'a> {
         rel: RelTag,
         nblocks: BlockNumber,
         ctx: &RequestContext,
-    ) -> anyhow::Result<()> {
-        anyhow::ensure!(rel.relnode != 0, "invalid relnode");
+    ) -> Result<(), RelationError> {
+        if rel.relnode == 0 {
+            return Err(RelationError::AlreadyExists);
```
[the remark this answers is not in the document]
Comment (awestover) [reply pageserver/src/pgdatadir_mapping.rs]: yes!

**Review by shanyp — APPROVED** [review]

File: pageserver/src/pgdatadir_mapping.rs [hunk 2]
```diff
@@ -875,32 +885,38 @@ impl<'a> DatadirModification<'a> {
         rel: RelTag,
         nblocks: BlockNumber,
         ctx: &RequestContext,
-    ) -> anyhow::Result<()> {
-        anyhow::ensure!(rel.relnode != 0, "invalid relnode");
+    ) -> Result<(), RelationError> {
+        if rel.relnode == 0 {
```
[the remark this answers is not in the document]
Comment (awestover) [reply pageserver/src/pgdatadir_mapping.rs]: yeah I changed it

**Comment by awestover** [comment]
getting this error message:
I'm pretty sure my edits here should not have affected the `remote_storage` tests though

ERROR test_runner/regress/test_remote_storage.py::test_remote_storage_upload_queue_retries[release-pg15-local_fs] - AssertionError: assert not ['2023-06-16T17:08:44.684097Z  WARN attach{tenant_id=32e6e4679ac3fdbba8b44c4d582d4b62}:load_remote_timeline{timeline_id=7428ce662630aef881d3dc85bbe698f9}:reconcile_with_remote: found future image layer 000000067F00000005000040000000000030-000000067F00000005000040000000000040__0000000001F27B88 on timeline 7428ce662630aef881d3dc85bbe698f9 remote_consistent_lsn is 0/1F27948\n', '2023-06-16T17:08:44.684150Z  WARN attach{tenant_id=32e6e4679ac3fdbba8b44c4d582d4b62}:load_remote_timeline{timeline_id=7428ce662630aef881d3dc85bbe698f9}:reconcile_with_remote: found future image layer 000000067F00000005000040050000000010-000000067F00000005000040050000000020__0000000001F27B88 on timeline 7428ce662630aef881d3dc85bbe698f9 remote_consistent_lsn is 0/1F27948\n', '2023-06-16T17:08:44.684231Z  WARN attach{tenant_id=32e6e4679ac3fdbba8b44c4d582d4b62}:load_remote_timeline{timeline_id=7428ce662630aef881d3dc85bbe698f9}:reconcile_with_remote: found future image layer 000000067F00000005000040050000000030-000000068000000000000004BD0000000001__0000000001F27B88 on timeline 7428ce662630aef881d3dc85bbe698f9 remote_consistent_lsn is 0/1F27948\n', '2023-06-16T17:08:44.684517Z  WARN attach{tenant_id=32e6e4679ac3fdbba8b44c4d582d4b62}:load_remote_timeline{timeline_id=7428ce662630aef881d3dc85bbe698f9}:reconcile_with_remote: found future image layer 000000067F00000005000040000000000040-000000067F00000005000040000000000050__0000000001F27B88 on timeline 7428ce662630aef881d3dc85bbe698f9 remote_consistent_lsn is 0/1F27948\n', '2023-06-16T17:08:44.684672Z  WARN attach{tenant_id=32e6e4679ac3fdbba8b44c4d582d4b62}:load_remote_timeline{timeline_id=7428ce662630aef881d3dc85bbe698f9}:reconcile_with_remote: found future image layer 000000067F00000005000004E70300000000-000000067F00000005000004EB000000000E__0000000001F27B88 on timeline 7428ce662630aef881d3dc85bbe698f9 remote_consistent_lsn is 0/1F27948\n', '2023-06-16T17:08:44.685183Z  WARN attach{tenant_id=32e6e4679ac3fdbba8b44c4d582d4b62}:load_remote_timeline{timeline_id=7428ce662630aef881d3dc85bbe698f9}:reconcile_with_remote: found future image layer 000000067F00000005000040040100000000-000000067F00000005000040050000000010__0000000001F27B88 on timeline 7428ce662630aef881d3dc85bbe698f9 remote_consistent_lsn is 0/1F27948\n', ...]

**Pull request merged**